modules/engine-prototype/src/main/scala/engine/Step.scala

```diff
@@ -0,0 +1,251 @@
+              both(configureTCS, configureInst).flatMap {
+              case (true, true) =>
+                  // XXX: Check running status
+                  Execution.observe.flatMap {
+                    case Some(fileId) => execute(m) // XXX: Shift current focus
+                    case None =>
+                      m.modify(
+                        Sequence.State.current.modify(s =>
+                          Failed.Standard(
+                            F2.Failed.Standard(
+                              s.core,
+                              tcsFailed = false,
+                              instFailed = false,
+                              observationFailed = true
+                            )
+                          )
+                        )
+                      ).void
+                  }
+
+              case (_, _) => F.pure(Unit)
+            }
+          }
+        }
+      }
+
+      sealed trait Ongoing extends Current
+      object Ongoing {
+        case class Standard(step: F2.Ongoing.Standard) extends Ongoing
+      }
+
+      sealed trait Failed extends Current
+      object Failed {
+        case class Standard(step: F2.Failed.Standard) extends Failed
+      }
+    }
+
+  }
+
+  sealed trait GMOS extends Step
+  case object GMOS
+
+  // This won't be accepted in cats-effect
+  // https://gist.github.com/djspiewak/a775b73804c581f4028fea2e98482b3c but with
+  // fs2 it should be possible, `parallelSequence` is already there.
+  private def both[F[_], A, B](fa: F[A], fb: F[B])(implicit F: Effect[F]): F[(A, B)] = ???
+
+}
+
+// current match {
+//   // No current Step
+//   case None => pending match {
+//     // No pending steps, done
+//     case Nil => m.setStatus(Status.Finished).void
+//     // More pending steps
+//     case (next :: remainder) => next.execute(m).flatMap {
+//       case Left(e) => m.setState(
+//         Sequence.State(
+//           F2(done, remainder, Some(Left(e))),
+//           Status.Failed
+//         )
+//       )
+//       case Right(d) => m.setSequence(
+//         F2((d :: done), remainder, None)
+//       ).flatMap(_.sequence.execute(m))
+//     }
+//   }
+//   case Some(_) => F.pure(Unit) // Event: Tried execute Sequence with an already ongoing step.
+// }
+// }
+
+
+//   // Alternative:
+//   // async.parallelSequence(List(Execution.configureTCS, Execution.configureInst("F2"))).flatMap(...)
+//   (for {
+//      t1 <- async.start(
+//        Execution.configureTCS
+//          //.flatMap(// x => sig.modify(Sequence.State.current.set(Some(Right(Ongoing())))) *> F.pure(x))
+//      )
+//      t2 <- async.start(Execution.configureInst("F2"))
+//      r1 <- t1
+//      r2 <- t2
+//    } yield (r1, r2)).flatMap {
+//     case (Right(_), Right(_)) =>
+//         Execution.observe.map {
+//           case Right(_) => ??? // Right(Done())
+//           case Left(_) => ??? // Left(Failed())
+//         }
+//     case _ => ??? // F.pure(Left(Failed()))
+//   }
+// }
```

Comment:
This is an experiment of concurrency in `fs2`. Notice now how the forking needs to be made explicit, in this case with `async.start`.

Something similar would have to be done for the unimplemented `both`, so that's why I left it commented out.